truffle/src/main/java/org/jruby/truffle/stdlib/DigestNodes.java

```diff
@@ -146,9 +148,17 @@ public UpdateNode(RubyContext context, SourceSection sourceSection) {
         @TruffleBoundary
         @Specialization(guards = "isRubyString(message)")
         public DynamicObject update(DynamicObject digestObject, DynamicObject message) {
-            final Rope rope = StringOperations.rope(message);
+            final MessageDigest digest = DigestLayoutImpl.INSTANCE.getDigest(digestObject);
+
+            RopeOperations.visitBytes(StringOperations.rope(message), new BytesVisitor() {
+
+                @Override
+                public void accept(byte[] bytes, int offset, int length) {
+                    digest.update(bytes, offset, length);
+                }
+
+            });
 
-            DigestLayoutImpl.INSTANCE.getDigest(digestObject).update(rope.getBytes(), rope.begin(), rope.byteLength());
```

Comment:
This is the whole point of this PR - don't allocate the full byte[] to output it (or feed it into a digest function in this case).